### DISCUSSION TITLE
Session event not processed correctly

### DISCUSSION THREAD

**Discussion by baymac** [WalletConnect/walletconnect-monorepo#3258] [Developer Support] [answered]
When the dApp is offline and wallet has some chain/account changes as follows:

1. Change Account -> B
2. Change Chain -> Y 
3. Change Account -> C
4. Change Chain -> Z 

Final Wallet state: Account C, Chain Z

When the dApp comes back online it processes requests in reverse order of their arrival (ref: https://github.com/WalletConnect/walletconnect-monorepo/blob/dc26b396c3a1a2f266143289842d7c7695eaf3ef/packages/sign-client/src/controllers/engine.ts#L917).

Final dApp state: Account B, Chain Y

I think this also unintuitive for the implementer, what is the reason the requests are processed in reverse order?

_Note: Above states assuming change chain doesn't chain account for simplicity._

**Comment by baymac** [comment]
I can outline another case where this causes problem:

Let's say dApp generates a session proposal with chains `[A, B]` in `requiredNamespace` and chains `[A, B, C]` in `optionalNamespaces`. The wallet has chain `C` selected.

Since Chain `C` is part of optional namespace, the wallet should add chain `C` to the session namespace and approve the proposal. Once the session is approved, then the wallet needs to send a `chainChanged` event to the dApp. Since session approval takes some time, both the requests go to the history of the client. When they are processed reverse chronologically, the `chainChanged` request is attempted to be processed first but there doesn't exist a session yet so it fails to execute the request and it cause other requests failure too.

A way to solve this is to add delay.

**Comment by baymac** [comment]
@ganchoradkov do you mind explaining the design choice behind processing request reverse chronologically?

**Reply by ganchoradkov** [reply, marked as answer]
Hey @baymac, the requests are processed in the order that they're received (first-in,first-out). This mechanism of the queue assures that all received requests are responded to, in the order they were received, unless they expire.

Regarding your second point about the `changeChanged` event and the wallet's selected chain, this behavior is starting to get deprecated and many wallets and dapps are moving to multi-chain approach i.e. all supported chains are active simultaneously. 
WalletConnect v2 employs this approach as well, and you're able to send requests to all approved chains without needing to switch first. It's up to the implementing wallet to figure out and handle its own behavior when and if at all, they wish to emit `chainChanged`

**Reply by baymac** [reply]
Hi @ganchoradkov, thank you for responding. During my testing, I've observed that the history is not handled with FIFO (First In, First Out) but rather with LIFO (Last In, First Out) logic. You can refer to the [comment](https://github.com/WalletConnect/walletconnect-monorepo/blob/dc26b396c3a1a2f266143289842d7c7695eaf3ef/packages/sign-client/src/controllers/engine.ts#L916) in the code, which also mentions this behaviour.

While I'm still working on finding explicit code evidence to support this, I have practically observed that the requests are indeed handled using LIFO. 

Here is the history of dApp:

In the below screenshot you can see that 1st history event has 3 accounts in `sessionUpdate`:

<img width="444" alt="Screenshot 2023-07-27 at 1 58 44 PM" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/23079344/cef13a89-59d9-419b-8ed4-eb044281a247">

And 2nd history event has 2 accounts in `sessionUpdate`:

<img width="441" alt="Screenshot 2023-07-27 at 1 58 52 PM" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/23079344/1c6ce6cb-1305-458f-807b-97fca2f209a7">

But in wallet history it was recorded in the reverse order.

**Reply by baymac** [reply]
This also applies to session event. The right side is wallet history and left side is dApp history. Note dApp was initially offline when 2 events occurred in wallet side.

<img width="1277" alt="Screenshot 2023-07-27 at 2 49 09 PM" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/23079344/d7f10eb7-ce1c-486f-96b8-581fc331685c">

**Reply by baymac** [reply]
@ganchoradkov sorry I am bugging you again. Do you have a response to this?
